### PR TITLE
store/tikv: copIteratorTaskSender is not initialized properly

### PR DESCRIPTION
**Pull request by tiancaiamao**
This bug is found by our recently added schrodinger leak test:

```
[case-id: leak-1525835901395] 2018/05/09 16:01:50 main.go:106: [fatal] goroutine seems to leak:
goroutine 742864 [select, 10 minutes]:
github.com/pingcap/tidb/store/tikv.(*copIteratorTaskSender).sendToTaskCh(0xc42dcb2680, 0xc4231da400, 0xc432ebbe00)
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/store/tikv/coprocessor.go:480 +0xdf
github.com/pingcap/tidb/store/tikv.(*copIteratorTaskSender).run(0xc42dcb2680)
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/store/tikv/coprocessor.go:450 +0x70
github.com/pingcap/tidb/store/tikv.(*copIteratorTaskSender).(github.com/pingcap/tidb/store/tikv.run)-fm()
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/store/tikv/coprocessor.go:444 +0x2a
github.com/pingcap/tidb/util/goroutine_pool.(*goroutine).workLoop(0xc427d82f60, 0xc4202da9c0)
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/util/goroutine_pool/gp.go:115 +0x171
created by github.com/pingcap/tidb/util/goroutine_pool.(*Pool).alloc
	/home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/util/goroutine_pool/gp.go:90 +0x97
```

It turns out to be that `copIteratorTaskSender` is not initialized properly, `finish` is nil and it hang forever here:

```
func (sender *copIteratorTaskSender) sendToTaskCh(t *copTask) (exit bool) {
	select {
	case sender.taskCh <- t:
	case <-sender.finished:
		exit = true
	}
	return
}
```

@coocood @zimulala

### CODE REVIEW

**Comment by shenli** [comment]
Please add a test case.

**Comment by tiancaiamao** [comment]
It's not easy to construct a case, schrodinger runs more than 5 hours to throw this error. @shenli 
Do you have any suggestion? @coocood

File: store/tikv/coprocessor.go [hunk 1]
```diff
@@ -436,9 +436,10 @@ func (it *copIterator) open(ctx context.Context) {
 		})
 	}
 	taskSender := &copIteratorTaskSender{
-		taskCh: taskCh,
-		wg:     &it.wg,
-		tasks:  it.tasks,
+		taskCh:   taskCh,
+		wg:       &it.wg,
+		tasks:    it.tasks,
+		finished: it.finished,
```
Comment (ngaut) [inline store/tikv/coprocessor.go]: The default value of finished is false, is there any chance the value is true?

Comment (tiancaiamao) [reply store/tikv/coprocessor.go]: `finished` is a `chan struct{}`, it should be initialized rather than `nil`

File: store/tikv/coprocessor.go [hunk 2]
```diff
@@ -770,7 +771,7 @@ func (worker *copIteratorWorker) calculateRemain(ranges *copRanges, split *copro
 
 func (it *copIterator) Close() error {
```
Comment (ngaut) [inline store/tikv/coprocessor.go]: Will Close method be called multiple times?

**Comment by coocood** [comment]
LGTM

File: store/tikv/coprocessor.go (hunk 2 above)
Comment (coocood) [reply store/tikv/coprocessor.go]: No, but the `it.finished` channel will be closed somewhere else.

**Pull request merged**